validator/index.js

```diff
@@ -25,50 +25,233 @@ const path = require('path');
 const program = require('commander');
 const vm = require('vm');
 const url = require('url');
+const util = require('util');
 
 /**
- * If the provided fileOrUrl start with 'http://' or 'https://', downloads
- * with the http or https module, otherwise reads the file with the file module.
- * Or, if the provided fileOrUrl is '-', reads from stdin.
- * Then, calls onSuccess with the data, or onFailure with the error message.
- * This method will check the HTTP status code as well and generate
- * brief error messages if it's not 200.
- * @param {!string} fileOrUrl
- * @param {!function(string)} onSuccess
- * @param {!function(string)} onFailure
+ * Convenience function to detect whether an argument is a ULR or
+ * (perhaps) a local file.
+ * @param {!string} url
+ * @returns {!boolean}
  */
-function readFileOrDownload(fileOrUrl, onSuccess, onFailure) {
-  const digestChunksFrom = function(response) {
+function isHttpOrHttpsUrl(url) {
+  return url.startsWith('http://') || url.startsWith('https://');
+}
+
+/**
+ * Creates a promise which reads from a file.
+ * @param {!string} name
+ * @returns {!Promise<(!string|!Error)>}
+ */
+function readFromFile(name) {
+  return new Promise(function(resolve, reject) {
+    fs.readFile(name, 'utf8', (err, data) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(data);
+      }
+    });
+  });
+}
+
+/**
+ * Creates a promise which reads from a stream.
+ * @param {!stream.Readable} readable
+ * @returns {!Promise<!string|Error>}
+ */
+function readFromReadable(readable) {
+  return new Promise(function(resolve, reject) {
     const chunks = [];
-    response.setEncoding('utf8');
-    response.on('data', function(chunk) { chunks.push(chunk); });
-    response.on('end', function() { onSuccess(chunks.join('')); });
-  };
-  if (fileOrUrl === '-') {
-    digestChunksFrom(process.stdin);
-    process.stdin.resume();
-    return;
-  }
-  if (fileOrUrl.startsWith('http://') || fileOrUrl.startsWith('https://')) {
-    const clientModule = fileOrUrl.startsWith('http://') ? http : https;
-    const req = clientModule.get(fileOrUrl, (response) => {
-      if (response.statusCode !== 200) {
-        onFailure('HTTP ' + response.statusCode);
-        return;
+    readable.setEncoding('utf8');
+    readable.on('data', (chunk) => { chunks.push(chunk); });
+    readable.on('end', () => { resolve(chunks.join('')); });
+    readable.on('error', (error) => { reject(error); });
+  });
+}
+
+/**
+ * Creates a promise which reads from standard input.
+ * @returns {!Promise<!string|Error>}
+ */
+function readFromStdin() {
+  return new Promise(function(resolve, reject) {
```

Comment:
I want to return a promise here, to make it easy to dispatch between readFromFile, readFromUrl, readFromStdin's results.